sys/kern/vm_map.c

```diff
@@ -416,19 +436,31 @@ vm_map_t *vm_map_clone(vm_map_t *map) {
   WITH_MTX_LOCK (&map->mtx) {
     vm_map_entry_t *it;
     TAILQ_FOREACH (it, &map->entries, link) {
-      vm_object_t *obj;
-      vm_map_entry_t *ent;
-
-      if (it->flags & VM_ENT_SHARED) {
-        vm_object_hold(it->object);
-        obj = it->object;
-      } else {
-        /* vm_object_clone will clone the data from the vm_object_t
-         * and will return the new object with ref_counter equal to one */
-        obj = vm_object_clone(it->object);
+      vm_map_entry_t *ent = vm_map_entry_copy(it);
+
+      if (!it->aref.amap && (it->flags & VM_ENT_SHARED)) {
+        /* We need to create amap, because we won't be able to share it if it
+         * is not created now. */
+        int slots = vaddr_to_slot(it->end - it->start);
+        it->aref.amap = vm_amap_alloc(slots + 16);
+        if (!it->aref.amap) {
+          return NULL;
+        }
+      }
+
+      if (it->aref.amap) {
+        if (it->flags & VM_ENT_SHARED) {
+          vm_amap_hold(it->aref.amap);
+          ent->aref = it->aref;
+        } else {
+          ent->aref.offset = 0;
+          ent->aref.amap = vm_amap_clone(it->aref);
+          if (!ent->aref.amap) {
+            klog("Unable to clone amap!");
+          }
```

Comment:
from this point - is ent a valid entry or not? because you ignore this  and always put ent into new map entries